File: app/services/analysis/capital.py
"""资金面分析引擎

提供主力资金流向、北向资金、融资融券、筹码分布分析
"""
import logging
from typing import Dict, Any, List
from datetime import datetime

from app.services.data.akshare_client import akshare_client
from app.services.data.tushare_client import get_tushare_client

logger = logging.getLogger(__name__)


class CapitalAnalyzer:
    """资金面分析器"""

    def analyze(self, code: str) -> Dict[str, Any]:
        """
        完整资金面分析
        
        Args:
            code: 股票代码
        
        Returns:
            资金面分析结果
        """
        code = code.strip()
        if '.' in code:
            code = code.split('.')[0]
        
        result = {
            'code': code,
            'analysis_time': datetime.now().strftime('%Y-%m-%d %H:%M:%S'),
            'money_flow': self._analyze_money_flow(code),
            'north_flow': self._analyze_north_flow(code),
            'margin': self._analyze_margin(code),
            'chip_distribution': self._analyze_chip_distribution(code),
            'capital_score': self._calc_capital_score(code),
        }
        
        return result

    def _analyze_money_flow(self, code: str) -> Dict[str, Any]:
        """分析主力资金流向"""
        flow_data = akshare_client.get_money_flow(code)
        
        if not flow_data:
            return {'available': False}
        
        # 判断主力动向
        main_net = flow_data.get('main_net_inflow', 0)
        trend = '流入' if main_net > 0 else '流出' if main_net < 0 else '平衡'
        
        # 计算主力净流入占比
        main_pct = flow_data.get('main_net_inflow_pct', 0)
        strength = '强' if abs(main_pct) > 5 else '中' if abs(main_pct) > 2 else '弱'
        
        return {
            'available': True,
            'date': flow_data.get('date'),
            'close': flow_data.get('close'),
            'change_pct': flow_data.get('change_pct'),
            'main_net_inflow': main_net,
            'main_net_inflow_pct': main_pct,
            'super_large_net': flow_data.get('super_large_net', 0),
            'large_net': flow_data.get('large_net', 0),
            'medium_net': flow_data.get('medium_net', 0),
            'small_net': flow_data.get('small_net', 0),
            'trend': trend,
            'strength': strength,
        }

    def _analyze_north_flow(self, code: str) -> Dict[str, Any]:
        """分析北向资金持股"""
        # 获取个股北向资金持股
        try:
            client = get_tushare_client()
            if not client.pro:
                return {'available': False}
            
            # 格式化代码
            ts_code = f"{code}.SH" if code.startswith('6') else f"{code}.SZ"
            
            # 获取沪深港通持股数据
            df = client.pro.hk_hold(ts_code=ts_code, start_date='', end_date='')
            
            if df is None or df.empty:
                return {'available': False}
            
            # 取最新数据
            latest = df.iloc[0]
            
            # 计算持股变化
            hold_ratio = latest.get('ratio', 0)
            
            return {
                'available': True,
                'date': str(latest.get('trade_date', '')),
                'hold_ratio': float(hold_ratio) if hold_ratio else 0,
                'hold_count': float(latest.get('count', 0)) if latest.get('count') else 0,
                'market': latest.get('market', ''),
            }
        except Exception as e:
            logger.warning(f"获取北向资金持股失败: {e}")
            return {'available': False}

    def _analyze_margin(self, code: str) -> Dict[str, Any]:
        """分析融资融券"""
        margin_data = akshare_client.get_margin_detail(code)
        
        if not margin_data:
            return {'available': False}
        
        # 计算融资融券趋势
        margin_balance = margin_data.get('融资余额', 0)
        margin_buy = margin_data.get('融资买入额', 0)
        margin_repay = margin_data.get('融资偿还额', 0)
        net_margin = margin_buy - margin_repay
        
        # 判断融资趋势
        trend = '增加' if net_margin > 0 else '减少' if net_margin < 0 else '平衡'
        
        return {
            'available': True,
            'date': margin_data.get('date'),
            'margin_balance': margin_balance,
            'margin_buy': margin_buy,
            'margin_repay': margin_repay,
            'net_margin': net_margin,
            'short_balance': margin_data.get('融券余额', 0),
            'trend': trend,
        }

    def _analyze_chip_distribution(self, code: str) -> Dict[str, Any]:
        """
        筹码分布估算
        
        基于成交量和价格波动估算获利盘/套牢盘比例
        """
        try:
            # 获取最近60天K线
            from app.services.data.stock_service import stock_service
            kline = stock_service.get_kline(code, period='daily', count=60)
            
            if not kline or len(kline) < 20:
                return {'available': False}
            
            # 当前价格
            current_price = kline[-1]['close']
            
            # 计算60日高低点
            high_60d = max(k['high'] for k in kline)
            low_60d = min(k['low'] for k in kline)
            
            # 估算获利盘比例（简化算法）
            # 假设筹码均匀分布，价格越高获利盘越多
            if high_60d > low_60d:
                profit_ratio = (current_price - low_60d) / (high_60d - low_60d) * 100
            else:
                profit_ratio = 50
            
            # 计算平均成本（成交量加权）
            total_amount = sum(k['amount'] for k in kline if k.get('amount'))
            total_volume = sum(k['vol'] for k in kline if k.get('vol'))
            avg_cost = total_amount / total_volume if total_volume > 0 else current_price
            
            # 判断筹码集中度
            price_range_pct = (high_60d - low_60d) / low_60d * 100
            concentration = '集中' if price_range_pct < 15 else '分散' if price_range_pct > 30 else '适中'
            
            return {
                'available': True,
                'current_price': current_price,
                'high_60d': high_60d,
                'low_60d': low_60d,
                'avg_cost': round(avg_cost, 2),
                'profit_ratio': round(profit_ratio, 2),
                'loss_ratio': round(100 - profit_ratio, 2),
                'concentration': concentration,
                'price_position': '高位' if current_price > avg_cost * 1.1 else '低位' if current_price < avg_cost * 0.9 else '中位',
            }
        except Exception as e:
            logger.warning(f"筹码分布分析失败: {e}")
            return {'available': False}

    def _calc_capital_score(self, code: str) -> Dict[str, Any]:
        """
        计算资金面评分
        
        评分维度：
        - 主力资金流向 (40分)
        - 北向资金持股 (30分)
        - 融资融券趋势 (30分)
        """
        score = 0
        details = {}
        
        # 1. 主力资金流向评分 (40分)
        flow_data = akshare_client.get_money_flow(code)
        if flow_data and flow_data.get('main_net_inflow') is not None:
            main_net = flow_data.get('main_net_inflow', 0)
            main_pct = flow_data.get('main_net_inflow_pct', 0)
            
            if main_net > 0:
                # 主力流入
                if main_pct > 5:
                    flow_score = 40
                elif main_pct > 2:
                    flow_score = 30
                elif main_pct > 0:
                    flow_score = 20
                else:
                    flow_score = 10
            else:
                # 主力流出
                if main_pct < -5:
                    flow_score = 0
                elif main_pct < -2:
                    flow_score = 5
                elif main_pct < 0:
                    flow_score = 10
                else:
                    flow_score = 15
            
            details['money_flow_score'] = flow_score
            details['money_flow_desc'] = f"主力净{'流入' if main_net > 0 else '流出'}{abs(main_net)/10000:.2f}万"
            score += flow_score
        
        # 2. 北向资金评分 (30分)
        try:
            client = get_tushare_client()
            if client.pro:
                ts_code = f"{code}.SH" if code.startswith('6') else f"{code}.SZ"
                df = client.pro.hk_hold(ts_code=ts_code, start_date='', end_date='')
                
                if df is not None and not df.empty:
                    hold_ratio = float(df.iloc[0].get('ratio', 0) or 0)
                    
                    # 北向资金持股比例评分
                    if hold_ratio > 5:
                        north_score = 30
                    elif hold_ratio > 3:
                        north_score = 25
                    elif hold_ratio > 1:
                        north_score = 20
                    elif hold_ratio > 0:
                        north_score = 10
                    else:
                        north_score = 0
                    
                    details['north_score'] = north_score
                    details['north_desc'] = f"北向持股{hold_ratio:.2f}%"
                    score += north_score
        except Exception as e:
            logger.warning(f"北向资金评分失败: {e}")
        
        # 3. 融资融券评分 (30分)
        margin_data = akshare_client.get_margin_detail(code)
        if margin_data:
            margin_balance = margin_data.get('融资余额', 0)
            net_margin = margin_data.get('融资买入额', 0) - margin_data.get('融资偿还额', 0)
            
            # 融资余额增加看多
            if net_margin > 0:
                margin_score = 25 if net_margin > 1000000 else 20 if net_margin > 0 else 15
            else:
                margin_score = 10 if net_margin > -1000000 else 5 if net_margin > -5000000 else 0
            
            details['margin_score'] = margin_score
            details['margin_desc'] = f"融资净{'买入' if net_margin > 0 else '偿还'}{abs(net_margin)/10000:.2f}万"
            score += margin_score
        
        # 评级
        if score >= 70:
            rating = '强势'
        elif score >= 50:
            rating = '偏多'
        elif score >= 30:
            rating = '中性'
        elif score >= 15:
            rating = '偏空'
        else:
            rating = '弱势'
        
        return {
            'total': score,
            'rating': rating,
            'details': details,
        }


# 全局实例
capital_analyzer = CapitalAnalyzer()

File: app/static/js/analysis/tabs/capital.js
/**
 * 资金面分析渲染模块
 * 
 * 职责:
 * - 渲染资金面分析内容（资金评分、筹码集中度、北向持股、资金流向、融资融券、筹码分布）
 */

/**
 * 渲染资金面分析内容
 * @param {Object} data - 资金面数据
 * @returns {string} HTML 字符串
 */
function renderCapital(data) {
    if (!data || data.error) return '<div class="text-center py-8 text-gray-400">暂无数据</div>';
    
    const moneyFlow = data.money_flow || {};
    const northFlow = data.north_flow || {};
    const margin = data.margin || {};
    const chipDist = data.chip_distribution || {};
    const capitalScore = data.capital_score || {};
    
    const fmt = (v, unit = '') => v != null ? (typeof v === 'number' ? v.toFixed(2) + unit : v) : '--';
    const fmtWan = (v) => {
        if (v == null) return '--';
        return (v / 10000).toFixed(2) + '万';
    };
    
    const flowColor = (v) => v > 0 ? 'text-rise' : v < 0 ? 'text-fall' : 'text-gray-600';
    
    return `
    <!-- 资金面评分 -->
    <div class="grid grid-cols-1 md:grid-cols-3 gap-4 mb-4">
        <div class="bg-gradient-to-br from-blue-50 to-blue-100 rounded-lg p-5 text-center">
            <h4 class="text-sm font-semibold text-blue-700 mb-2">资金面评分</h4>
            <div class="text-2xl sm:text-3xl font-bold text-blue-900">${fmt(capitalScore.total)}</div>
            <div class="text-xs sm:text-sm text-blue-600 mt-1">${capitalScore.rating || '--'}</div>
        </div>
        <div class="bg-gradient-to-br from-green-50 to-green-100 rounded-lg p-5 text-center">
            <h4 class="text-sm font-semibold text-green-700 mb-2">筹码集中度</h4>
            <div class="text-xl sm:text-2xl font-bold text-green-900">${chipDist.concentration || '--'}</div>
            <div class="text-xs sm:text-sm text-green-600 mt-1">${chipDist.price_position || '--'}</div>
        </div>
        <div class="bg-gradient-to-br from-purple-50 to-purple-100 rounded-lg p-5 text-center">
            <h4 class="text-sm font-semibold text-purple-700 mb-2">北向持股</h4>
            <div class="text-xl sm:text-2xl font-bold text-purple-900">${fmt(northFlow.hold_ratio, '%')}</div>
            <div class="text-xs sm:text-sm text-purple-600 mt-1">${northFlow.date ? northFlow.date.slice(4, 6) + '-' + northFlow.date.slice(6, 8) : '--'}</div>
        </div>
    </div>

    <!-- 资金流向 + 融资融券 + 筹码分布 -->
    <div class="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-3 gap-4">
        <!-- 主力资金流向 -->
        ${moneyFlow.available ? `
        <div class="bg-gray-50 rounded-lg p-4">
            <h4 class="text-sm font-semibold text-gray-600 mb-3">💰 主力资金流向 <span class="${flowColor(moneyFlow.main_net_inflow)} font-normal">(${moneyFlow.trend || '--'})</span></h4>
            <div class="space-y-2 text-xs sm:text-sm">
                <div class="flex justify-between items-center"><span class="text-gray-500">主力净流入</span><span class="font-semibold tabular-nums ${flowColor(moneyFlow.main_net_inflow)}">${fmtWan(moneyFlow.main_net_inflow)}</span></div>
                <div class="flex justify-between items-center"><span class="text-gray-500">主力占比</span><span class="font-semibold tabular-nums">${fmt(moneyFlow.main_net_inflow_pct, '%')}</span></div>
                <div class="border-t pt-2 mt-2">
                    <div class="flex justify-between items-center"><span class="text-gray-500">超大单</span><span class="font-semibold tabular-nums ${flowColor(moneyFlow.super_large_net)}">${fmtWan(moneyFlow.super_large_net)}</span></div>
                    <div class="flex justify-between items-center"><span class="text-gray-500">大单</span><span class="font-semibold tabular-nums ${flowColor(moneyFlow.large_net)}">${fmtWan(moneyFlow.large_net)}</span></div>
                    <div class="flex justify-between items-center"><span class="text-gray-500">中单</span><span class="font-semibold tabular-nums ${flowColor(moneyFlow.medium_net)}">${fmtWan(moneyFlow.medium_net)}</span></div>
                    <div class="flex justify-between items-center"><span class="text-gray-500">小单</span><span class="font-semibold tabular-nums ${flowColor(moneyFlow.small_net)}">${fmtWan(moneyFlow.small_net)}</span></div>
                </div>
            </div>
            <div class="mt-3 pt-2 border-t text-[10px] text-gray-400 leading-relaxed">
                <span class="font-medium text-gray-500">📋 统计规则(Tushare):</span> 小单&lt;5万 | 中单5~20万 | 大单20~100万 | 特大单≥100万 | 主力=大单+特大单
            </div>
        </div>
        ` : `
        <div class="bg-gray-50 rounded-lg p-4">
            <h4 class="text-sm font-semibold text-gray-600 mb-3">💰 主力资金流向</h4>
            <div class="text-center py-8 text-gray-400 text-xs sm:text-sm">数据源连接失败<br>请稍后重试</div>
        </div>
        `}

        <!-- 融资融券 -->
        ${margin.available ? `
        <div class="bg-gray-50 rounded-lg p-4">
            <h4 class="text-sm font-semibold text-gray-600 mb-3">📊 融资融券 <span class="${margin.net_margin > 0 ? 'text-rise' : 'text-fall'} font-normal">(${margin.trend || '--'})</span></h4>
            <div class="space-y-2 text-xs sm:text-sm">
                <div class="flex justify-between items-center"><span class="text-gray-500">融资余额</span><span class="font-semibold tabular-nums">${fmtWan(margin.margin_balance)}</span></div>
                <div class="flex justify-between items-center"><span class="text-gray-500">融资买入</span><span class="font-semibold tabular-nums">${fmtWan(margin.margin_buy)}</span></div>
                <div class="flex justify-between items-center"><span class="text-gray-500">融资偿还</span><span class="font-semibold tabular-nums">${fmtWan(margin.margin_repay)}</span></div>
                <div class="flex justify-between items-center border-t pt-2 mt-2"><span class="text-gray-500">融资净买入</span><span class="font-semibold tabular-nums ${flowColor(margin.net_margin)}">${fmtWan(margin.net_margin)}</span></div>
                <div class="flex justify-between items-center"><span class="text-gray-500">融券余额</span><span class="font-semibold tabular-nums">${fmtWan(margin.short_balance)}</span></div>
            </div>
        </div>
        ` : `
        <div class="bg-gray-50 rounded-lg p-4">
            <h4 class="text-sm font-semibold text-gray-600 mb-3">📊 融资融券</h4>
            <div class="text-center py-8 text-gray-400 text-xs sm:text-sm">暂无数据</div>
        </div>
        `}

        <!-- 筹码分布 -->
        ${chipDist.available ? `
        <div class="bg-gray-50 rounded-lg p-4">
            <h4 class="text-sm font-semibold text-gray-600 mb-3">🎯 筹码分布</h4>
            <div class="space-y-2 text-xs sm:text-sm">
                <div class="flex justify-between items-center"><span class="text-gray-500">当前价格</span><span class="font-semibold tabular-nums">${fmt(chipDist.current_price)}</span></div>
                <div class="flex justify-between items-center"><span class="text-gray-500">60日高点</span><span class="font-semibold tabular-nums text-rise">${fmt(chipDist.high_60d)}</span></div>
                <div class="flex justify-between items-center"><span class="text-gray-500">60日低点</span><span class="font-semibold tabular-nums text-fall">${fmt(chipDist.low_60d)}</span></div>
                <div class="flex justify-between items-center"><span class="text-gray-500">平均成本</span><span class="font-semibold tabular-nums">${fmt(chipDist.avg_cost)}</span></div>
                <div class="border-t pt-2 mt-2">
                    <div class="flex justify-between items-center"><span class="text-gray-500">获利盘</span><span class="font-semibold tabular-nums text-rise">${fmt(chipDist.profit_ratio, '%')}</span></div>
                    <div class="flex justify-between items-center"><span class="text-gray-500">套牢盘</span><span class="font-semibold tabular-nums text-fall">${fmt(chipDist.loss_ratio, '%')}</span></div>
                </div>
            </div>
        </div>
        ` : `
        <div class="bg-gray-50 rounded-lg p-4">
            <h4 class="text-sm font-semibold text-gray-600 mb-3">🎯 筹码分布</h4>
            <div class="text-center py-8 text-gray-400 text-xs sm:text-sm">暂无数据</div>
        </div>
        `}
    </div>
    `;
}
